### DISCUSSION TITLE
Controller Model Injection interference with route variable injection?

### DISCUSSION THREAD

**Discussion by Korko** [laravel/framework#35689] [Q&A] [answered]
- Laravel Version: 7.29.3
- PHP Version: 7.4.13
- Database Driver & Version: N/A

### Description:
Hi everyone,
I was adding a new route which combine a model injection (with column name) and some other variable but in my controller the variable doesn't seem to be detected. Maybe I'm wrong but I thought in the controller method, depending on the parameter name, it should get the according route variable. In my case, I have 2 parameters in the route but in the controller I have to put 3 in order to get the last one.

### Steps To Reproduce:
In my routes/web.php, I have this route:
`Route::get('/email/{email:hash}/{version}', [EmailController::class, 'updateStatus'])->name('pixel')->middleware('signed');`

And my controller:
```
<?php

namespace App\Http\Controllers;

use App\Models\Mail as MailModel;

class EmailController extends Controller
{
    public function updateStatus(MailModel $mail, $version, $test)
    {
        dd($mail, $version, $test);
    }
}
```

What I expected:
$mail is a MailModel, $version is "0" and $test is null (or "GyeDG", I don't care for this parameter)

What I get:
$mail is a MailModel, $version is the hash ("GyeDG") and $test is the version ("0")

So the parameters are always filled in order, not by using reflection on the parameter name to try to guess the actual route variable? As I don't care for the hash (already got it with the model injection), I still have to fill the parameter in my controller even if I won't use it?

Thanks

**Comment by kohlerdominik** [comment, marked as answer]
Hi

From what i can see there's no parameter-name-mapping [here](https://github.com/laravel/framework/blob/130168c9dcd399f6b42bccfe4882b88c81a4edfe/src/Illuminate/Routing/RouteDependencyResolverTrait.php#L56). Not-type-hinted variables are filled in the order the parameters are set in the route.

I think, a much better approach in general is using the Request-instance rather than injecting values in the controller. This would work like this:
```php
<?php

namespace App\Http\Controllers;

use App\Models\Mail as MailModel;

class EmailController extends Controller
{
    public function updateStatus(MailModel $mail, Request $request)
    {
        $version = $request->route('version');
    }
}
```

**Reply by Korko** [reply]
I like your approach. I think it would be interesting to have the parameter-name-mapping though.
